### PR TITLE
Before murder only once

### PR DESCRIPTION
**Pull request by shivnagarajan**
https://github.com/camilo/unicorn/pull/5

Basically diff to our shopify/unicorn branch

@fbogsany @daniellaniyo

### CODE REVIEW

**Comment by daniellaniyo** [comment]
👍

File: lib/unicorn/http_server.rb [hunk 1]
```diff
@@ -91,6 +92,7 @@ def initialize(app, options = {})
     @self_pipe = []
     @workers = {} # hash maps PIDs to Workers
     @sig_queue = [] # signal queue used for self-piping
+    @before_murder_called = Set.new() # ensures before_murder is called only once
```
Comment (fbogsany) [inline lib/unicorn/http_server.rb]: Parens not necessary.

File: lib/unicorn/http_server.rb [hunk 2]
```diff
@@ -475,7 +478,10 @@ def murder_lazy_workers
       next_sleep = 0
       logger.error "worker=#{worker.nr} PID:#{wpid} timeout " \
                    "(#{diff}s > #{@timeout}s), killing"
-      before_murder.call(self, worker, wpid) if before_murder
+      if before_murder && !@before_murder_called.include?(wpid)
+        before_murder.call(self, worker, wpid)
+        @before_murder_called.add(wpid) # This ensures we call before_murder only once
+      end
```
Comment (fbogsany) [inline lib/unicorn/http_server.rb]: Could write this as:
```ruby
before_murder_call(self, worker, wpid) if before_murder and @before_murder_called.add?(wpid)
```
The `.add?` call adds the `wpid` to the `Set` and returns `nil` if it was already there or `self` if it wasn't.

Not sure if that's too much code :golf:

File: t/t0024-before-murder_once.sh [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    \$stderr.puts "kill_worker #{signal} #{wpid} #{@iteration}"
+    if @iteration > 0 then
+      Process.kill(signal, wpid)
+    end
+    @iteration = @iteration + 1
+  rescue Errno::ESRCH
+    @before_murder_called.delete(wpid)
+    worker = @workers.delete(wpid) and worker.close rescue nil
+  end
+end
```
Comment (fbogsany) [inline t/t0024-before-murder_once.sh]: Might be nicer to use `alias_method` here, so:
```ruby
class Unicorn::HttpServer
  alias_method :kill_worker_always, :kill_worker
  def kill_worker(signal, wpid)
    @iteration ||= 0
    \$stderr.puts "kill_worker #{signal} #{wpid} #{@iteration}"
    kill_worker_always(signal, wpid) if @iteration > 0
    @iteration += 1
  end
end
```
This ensures we don't need to update the test if the implementation of `kill_worker` changes.

**Comment by shivnagarajan** [comment]
Adding Camilo, this allows us to ship https://github.com/Shopify/shopify/issues/78263.
I've bumped the version and uploaded the gem to packagecloud.

@camilo

File: unicorn.gemspec [hunk 4]
```diff
@@ -14,7 +14,7 @@ end.compact
 
 Gem::Specification.new do |s|
   s.name = %q{unicorn-camilo}
```
Comment (camilo) [inline unicorn.gemspec]: we can also call this bad boy `unicorn-shopify` at this point I think

**Comment by shivnagarajan** [comment]
new name, and uploaded to package_cloud

**Pull request merged**